ambry-store/src/main/java/com.github.ambry.store/IndexValue.java

```diff
@@ -13,114 +13,323 @@
+    return expiresAtMs;
   }
 
+  /**
+   * @return the original message offset of the {@link IndexValue}
+   */
   long getOriginalMessageOffset() {
-    return value.getLong(Blob_Size_In_Bytes + Offset_Size_In_Bytes + Flag_Size_In_Bytes + Expires_At_Ms_Size_In_Bytes);
+    return originalMessageOffset;
   }
 
+  /**
+   * @return the operation time in ms of the index value
+   */
+  long getOperationTimeInMs() {
+    return operationTimeInMs;
+  }
+
+  /**
+   * @return the serviceId of the {@link IndexValue}
+   */
+  short getServiceId() {
+    return serviceId;
+  }
+
+  /**
+   * @return the containerId of the {@link IndexValue}
+   */
+  short getContainerId() {
+    return containerId;
+  }
+
+  /**
+   * Sets the {@link Flags} for the {@link IndexValue}
+   * @param flag the {@link Flags} that needs to be set in the {@link IndexValue}
+   */
   void setFlag(Flags flag) {
-    value.put(Blob_Size_In_Bytes + Offset_Size_In_Bytes, (byte) (getFlags() | (1 << flag.ordinal())));
+    flags = (byte) (flags | (1 << flag.ordinal()));
   }
 
+  /**
+   * Updates the {@link Offset} of the {@link IndexValue}
+   * @param newOffset the new {@link Offset} to be updated for the {@link IndexValue}
+   */
   void setNewOffset(Offset newOffset) {
-    long originalMessageOffset = offset.getName().equals(newOffset.getName()) ? offset.getOffset() : -1;
+    originalMessageOffset = offset.getName().equals(newOffset.getName()) ? offset.getOffset() : -1;
     offset = newOffset;
-    value.putLong(Blob_Size_In_Bytes, offset.getOffset());
-    value.putLong(Blob_Size_In_Bytes + Offset_Size_In_Bytes + Flag_Size_In_Bytes + Expires_At_Ms_Size_In_Bytes,
-        originalMessageOffset);
   }
 
   void clearOriginalMessageOffset() {
-    value.putLong(Blob_Size_In_Bytes + Offset_Size_In_Bytes + Flag_Size_In_Bytes + Expires_At_Ms_Size_In_Bytes, -1);
+    originalMessageOffset = -1;
   }
 
+  /**
+   * Sets the size of the {@link IndexValue}
+   * @param size the size that needs to be set for the {@link IndexValue}
+   */
   void setNewSize(long size) {
-    value.putLong(0, size);
+    this.size = size;
   }
 
+  /**
+   * @return the {@link ByteBuffer} representation of this {@link IndexValue}
+   */
   ByteBuffer getBytes() {
+    ByteBuffer value;
+    switch (version) {
+      case PersistentIndex.VERSION_0:
+        value = ByteBuffer.allocate(INDEX_VALUE_SIZE_IN_BYTES_V0);
+        value.putLong(size);
+        value.putLong(offset.getOffset());
+        value.put(flags);
+        value.putLong(expiresAtMs);
+        value.putLong(originalMessageOffset);
+        value.position(0);
+        break;
+      case PersistentIndex.VERSION_1:
+        value = ByteBuffer.allocate(INDEX_VALUE_SIZE_IN_BYTES_V1);
+        value.putLong(size);
+        value.putLong(offset.getOffset());
+        value.put(flags);
+        value.putInt(expiresAtMs != Utils.Infinite_Time ? (int) (expiresAtMs / Time.MsPerSec) : (int) expiresAtMs);
```

Comment:
since you calculate the sec value in the constructors, why not just store it as a class variable instead of this calculation?